main/contracts/index.ts

```diff
@@ -44,44 +85,89 @@ function parseAbi (abiData: string): Interface | undefined {
   }
 }
 
-// function scanEndpoint (contractAddress: Address, chainId: string) {
-//   if (chainId === '0x1') {
-//     return `https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`
-//   } else if (chainId === '0x89') {
-//     return `https://api.polygonscan.com/api?module=contract&action=getsourcecode&address=0x0d500B1d8E8eF31E21C99d1Db9A6444d3ADf1270&apikey=YourApiKeyToken`
-//   } else if ('optimism') {
-//     return `https://api-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=0x80AA7cb0006d5DDD91cce684229Ac6e398864606&apikey=YourApiKeyToken`
-//   } else if ('arbitrum') {
-//     return `https://api.arbiscan.io/api?module=contract&action=getsourcecode&address=0x0000000000000000000000000000000000001004&apikey=YourApiKeyToken`
-//   }
-// }
+const scanEndpointMap = {
+  '0x1': {
+    name: 'etherscan',
+    url: (contractAddress: Address) => `https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`,
+  },
+  '0x89': {
+    name: 'polygonscan',
+    url: (contractAddress: Address) => `https://api.polygonscan.com/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=2P3U9T63MT26T1X64AAE368UNTS9RKEEBB`,
+  },
+  '0xa': {
+    name: 'optimistic.etherscan',
+    url: (contractAddress: Address) => `https://api-optimistic.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`,
+  },
+  '0xa4b1': {
+    name: 'arbiscan',
+    url: (contractAddress: Address) => `https://api.arbiscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=VP126CP67QVH9ZEKAZT1UZ751VZ6ZTIZAD`
+  }
+}
 
-async function fetchSourceCode (contractAddress: Address, chainId: string) {
-  const res = await fetch(`https://api.etherscan.io/api?module=contract&action=getsourcecode&address=${contractAddress}&apikey=3SYU5MW5QK8RPCJV1XVICHWKT774993S24`)
+function sourcifyEndpoint (contractAddress: Address, chainId: string) {
+  return `https://sourcify.dev/server/files/any/${hexToNumberString(chainId)}/${contractAddress}`
+}
 
-  if (res.status === 200 && (res.headers.get('content-type') || '').toLowerCase().includes('json')) {
-    const parsedResponse = await (res.json() as Promise<EtherscanSourceCodeResponse>)
+async function parseResponse <T>(response: Response): Promise<T | undefined> {
+  if (response?.status !== 200 || !(response?.headers.get('content-type') || '').toLowerCase().includes('json')) {
+    return Promise.resolve(undefined)
```

Comment:
don't need to explicitly return this, can flip the check around and only return the json if its true

Reply:
Not sure how inverting the logic helps?  We still need to resolve undefined